modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -148,67 +176,48 @@ object Sink {
-                 s"Waiting for ${retry.cumulativeDelay} and retrying to write ${records.size} records to ${config.streamName}"
-               ) *>
-                 Timer[F].sleep(retry.cumulativeDelay)
-           }
-      putRecordsResult <- withRetry
-      failuresRetried <- if (putRecordsResult.getFailedRecordCount != 0) {
-                           val failurePairs = records.zip(putRecordsResult.getRecords.asScala).filter(_._2.getErrorMessage != null)
-                           val (failedRecords, failedResults) = failurePairs.unzip
-                           val logging = logErrorsSummary(getErrorsSummary(failedResults))
-                           val maybeRetrying =
-                             for {
-                               nextRetry <- retryPolicy.decideNextRetry(retryStatus)
-                               maybeRetried <- nextRetry match {
-                                                 case PolicyDecision.DelayAndRetry(delay) =>
-                                                   writeToKinesis(
-                                                     blocker,
-                                                     config,
-                                                     kinesis,
-                                                     failedRecords,
-                                                     retryPolicy,
-                                                     retryStatus.addRetry(delay)
-                                                   )
-                                                 case PolicyDecision.GiveUp =>
-                                                   Sync[F].raiseError[Unit](
-                                                     new RuntimeException(
-                                                       s"Maximum number of retries reached for ${failedRecords.size} records"
-                                                     )
-                                                   )
-                                               }
-                             } yield maybeRetried
-                           logging *> maybeRetrying
-                         } else
-                           Sync[F].unit
-    } yield failuresRetried
-  }
+    retryPolicy: RetryPolicy[F]
+  ): F[List[PutRecordsRequestEntry]] =
+    Logger[F].debug(s"Writing ${records.size} records to ${config.streamName}") *>
+      blocker
+        .blockOn(Sync[F].delay(putRecords(kinesis, config.streamName, records)))
+        .retryingOnFailuresAndAllErrors(
+          policy = retryPolicy,
+          wasSuccessful = { result =>
+            // It's a success if _any_ record has a status which we want to handle ourselves.
+            // This includes if we got throttled by Kinesis.
+            result.getFailedRecordCount.toInt === 0 || result.getRecords.asScala
+              .flatMap(record => Option(record.getErrorCode))
+              .exists(_ === "ProvisionedThroughputExceededException")
```

Comment:
I disagree with retrying internal failures indefinitely.  And my reasons are... if we retry indefinitely, and if there is ever a problem with our kinesis records, then we never discover about it.  The app would never crash, we would never get an exception in sentry to say what happened.  If you think about it, why do we have a `maxRetries` config option at all?  It's because there are circumstances where it is better to let the app crash.  It does a hard reset on any JVM resources (in case the client was the problem), and we get a notification about it.

Reply:
I think that there can't be `InternalFailure`s due to our Kinesis records, they are just byte arrays, but I agree with you that in case of AWS outage for instance we might get them for a long time and even if crashing and restarting the app won't change anything, at least we get notified